scripts/google_replicate.py

```diff
@@ -0,0 +1,240 @@
+    """
+    client = storage.Client()
+    bucket = client.bucket(bucket_name)
+    return bucket.exists()
+
+
+def blob_exists(bucket_name, blob_name):
+    """
+    check if blob exists or not!
+    """
+    if bucket_exists(bucket_name):
+        client = storage.Client()
+        bucket = client.bucket(bucket_name)
+        blob = Blob(blob_name, bucket)
+        return blob.exists()
+    return False
+
+
+def check_blob_name_exists_and_match_md5_size(sess, bucket_name, blob_name, fi):
+    """
+    check if blob object exists or not
+    Args:
+        bucket_name(str): the name of bucket
+        blob_name(str): the name of blob
+        fi(dict): file info dictionary
+    Returns:
+        bool: indicating value if the blob is exist or not
+    """
+    url = "https://www.googleapis.com/storage/v1/b/{}/o/{}".format(
+        bucket_name, urllib.quote(blob_name, safe="")
+    )
+    res = sess.request(method="GET", url=url)
+    return (
+        res.status_code == 200
+        and res.json["size"] == fi.get("size")
+        and base64.b64decode(res.json()["md5Hash"]).encode("hex") == fi.get("md5")
+    )
+
+
+def update_indexd(fi):
+    """
+    update a record to indexd
+    Args:
+        fi(dict): file info
+    Returns:
+         None
+    """
+    gs_bucket_name = get_bucket_name(fi, PROJECT_MAP)
+    gs_object_name = "{}/{}".format(fi.get("id"), fi.get("filename"))
+
+    try:
+        doc = indexclient.get(fi.get("id", ""))
+        if doc is not None:
+            url = "gs://{}/{}".format(gs_bucket_name, gs_object_name)
+            if url not in doc.urls:
+                doc.urls.append(url)
+                doc.patch()
+            return
+    except Exception as e:
+        raise APIError(
+            "INDEX_CLIENT: Can not update the record with uuid {}. Detail {}".format(
+                fi.get("id", ""), e
+            )
+        )
+
+    urls = ["https://api.gdc.cancer.gov/data/{}".format(fi["fileid"])]
+
+    if blob_exists(gs_bucket_name, gs_object_name):
+        urls.append("gs://{}/{}".format(gs_bucket_name, gs_object_name))
+
+    try:
+        doc = indexclient.create(
+            did=fi.get("id", ""),
+            hashes={"md5": fi.get("md5", "")},
+            size=fi.get("size", 0),
+            urls=urls,
+        )
+        if doc is None:
+            raise APIError(
+                "INDEX_CLIENT: Fail to create a record with uuid {}".format(
+                    fi.get("id", "")
+                )
+            )
+    except Exception as e:
+        raise APIError(
+            "INDEX_CLIENT: Can not create the record with uuid {}. Detail {}".format(
+                fi.get("id", ""), e.message
+            )
+        )
+
+
+def exec_google_copy(fi, global_config):
+    """
+    copy a file to google bucket.
+    Args:
+        fi(dict): a dictionary of a copying file
+        global_config(dict): a configuration
+    Returns:
+        DataFlowLog
+    """
+    client = storage.Client()
```

Comment:
so, this is a huge change, but probably all of this Google logic should actually live in the `cirrus` library, as that is what its intention was, to wrap all this Google stuff so we can reuse elsewhere.

It handles setting up authorized session and everything, but I'm beginning to think that long-term we should really migrate a *lot* of this code to `cirrus`...

Reply:
1) to allow for a single place to handle our interface to Google's API should they change things
2) for modularized code that can be imported and used in other future software that may need to do some of these general actions like copying to a bucket